wireapp/kalium
# refactor: optimise paginated conversation list queries [WPB-11808]

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-11808" title="WPB-11808" target="_blank"><img alt="Bug" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10803?size=medium" />WPB-11808</a>  [Android] app stalls when fetching a page of conversations from the local DB
  </td></table>
  <br />
 










----
#### PR Submission Checklist for internal contributors

- The **PR Title**

- The **PR Description**
----

# What's new in this PR?

### Issues

The app stalls when fetching a page of conversations from the local DB. 

### Causes (Optional)

Our queries for fetching conversation list pages are very complex, many joins with different tables resulting in full table scans, especially Message table.

### Solutions

Optimised queries, joins made directly instead of using views so that it doesn't make full scans but just searches using existing indexes. Also rearranged joins to get self user so that it's executed once and not 4 times
Removed all unnecessary joins from query that only counts elements.
Moved `ConversationDetails` and `ConversationDetailsWithEvents` views to dedicated files along with all queries related to them and unify naming of functions.
Making a search with wildcard character % at the beginning results in full table scan of conversations (https://www.sqlite.org/optoverview.html#the_like_optimization) so now it's separated by creating two functions: for searching conversations with query and just fetching without searching, so that it doesn't scan if it's not actually necessary.

### Testing

#### Test Coverage (Optional)


#### How to Test

Check how query is executed by calling `EXPLAIN QUERY PLAN...` or just open the app and scroll the conversations.

### Attachments (Optional)

#### Query before:
![query_before](https://github.com/user-attachments/assets/89fcc983-566d-4857-9620-bacc479f7d5b)

#### Query after:
![query_after](https://github.com/user-attachments/assets/ed5616d2-0de8-4294-b12b-bcad187808ba)


----
#### PR Post Submission Checklist for internal contributors (Optional)

----
#### PR Post Merge Checklist for internal contributors

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.
